src/accountsdb/buffer_pool.zig

```diff
@@ -0,0 +1,1175 @@
+        self: *BufferPool,
+        f_idx: FrameIndex,
+        file_id: FileId,
+        frame_aligned_file_offset: FileOffset,
+    ) error{CannotOverwriteAliveInfo}!void {
+        std.debug.assert(frame_aligned_file_offset % FRAME_SIZE == 0);
+
+        if (self.frames_metadata.rc[f_idx].isAlive()) {
+            // not-found indices should always have 0 active readers
+            return error.CannotOverwriteAliveInfo;
+        }
+
+        self.frames_metadata.freq[f_idx] = 0;
+        self.frames_metadata.in_queue[f_idx] = .none;
+        self.frames_metadata.rc[f_idx].reset();
+
+        self.frames_metadata.key[f_idx] = .{
+            .file_id = file_id,
+            .file_offset = frame_aligned_file_offset,
+        };
+
+        const key: FileIdFileOffset = .{
+            .file_id = file_id,
+            .file_offset = frame_aligned_file_offset,
+        };
+
+        self.frame_map.putAssumeCapacityNoClobber(key, f_idx);
+    }
+
+    /// Frames with an associated rc of 0 are up for eviction, and which frames
+    /// are evicted first is up to the LFU.
+    fn evictUnusedFrame(self: *BufferPool) error{CannotResetAlive}!void {
+        const evicted = self.eviction_lfu.evict(self.frames_metadata);
+        self.free_list.appendAssumeCapacity(evicted);
+        const did_remove = self.frame_map.remove(self.frames_metadata.key[evicted]);
+        if (!did_remove) {
+            std.debug.panic(
+                "evicted a frame that did not exist in frame_map, frame: {}\n",
+                .{evicted},
+            );
+        }
+        @memset(&self.frames[evicted], 0xAA);
+        try self.frames_metadata.resetFrame(evicted);
+    }
+
+    pub fn read(
+        self: *BufferPool,
+        /// used for temp allocations, and the returned .indices slice
+        allocator: std.mem.Allocator,
+        file: std.fs.File,
+        file_id: FileId,
+        /// inclusive
+        file_offset_start: FileOffset,
+        /// exclusive
+        file_offset_end: FileOffset,
+    ) !CachedRead {
+        return if (use_io_uring)
+            self.readIoUringSubmitAndWait(
+                allocator,
+                file,
+                file_id,
+                file_offset_start,
+                file_offset_end,
+            )
+        else
+            self.readBlocking(
+                allocator,
+                file,
+                file_id,
+                file_offset_start,
+                file_offset_end,
+            );
+    }
+
+    fn readIoUringSubmitAndWait(
+        self: *BufferPool,
+        /// used for temp allocations, and the returned .indices slice
+        allocator: std.mem.Allocator,
+        file: std.fs.File,
+        file_id: FileId,
+        /// inclusive
+        file_offset_start: FileOffset,
+        /// exclusive
+        file_offset_end: FileOffset,
+    ) !CachedRead {
+        if (builtin.os.tag != .linux)
+            @compileError("io_uring only available on linux - unsupported target");
```

Comment:
Maybe use the `use_io_uring` here instead, so that's synced.

Reply:
fixed in b3aa41d4c7a8e90beeeed1d917bbf2b2c374183c

Reply:
I think this is good as-is if someone calls readIoUring explicitly from linux. Most call sites should just use the read method.